File: src/main/java/discretemath/structure/function/concrete/FloorFunction.java
package discretemath.structure.function.concrete;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;

import discretemath.structure.function.CalculableFunction;
import discretemath.structure.set.Sets;

public final class FloorFunction extends CalculableFunction<BigDecimal, BigInteger> {

	public FloorFunction() {
		super(Sets.R, Sets.Z);
	}

	@Override
	public boolean isOnto() {
		return true;
	}

	@Override
	public boolean isOneToOne() {
		return false;
	}

	@Override
	protected BigInteger calculate(BigDecimal value) {
		return value.setScale(0, RoundingMode.FLOOR).toBigInteger();
	}
}

File: src/main/scala/discretemath/structure/function/concrete/FloorFunction.scala
package discretemath.structure.function.concrete

import discretemath.structure.function.CalculableFunction
import discretemath.structure.set.Sets

import scala.math.{BigDecimal, BigInt}

final class FloorFunction() extends CalculableFunction[BigDecimal, BigInt](Sets.R, Sets.Z) {

  override def isOnto = true

  override def isOneToOne = false

  override def calculate(value: BigDecimal): BigInt = value.setScale(0, BigDecimal.RoundingMode.FLOOR).toBigInt
}
